### ISSUE TITLE
dialer is closed error

### ISSUE THREAD

**Issue by jSierraB3991** [GoogleCloudPlatform/cloud-sql-go-connector#777]
### Bug Description

Sure, here's a draft for your GitHub issue:

---

**Issue Summary:**
After updating a Go dependency yesterday, the application utilizing `cloud-sql-go-connector` began rejecting database connections.

**Details:**
![image](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/assets/30013840/fc7f8431-4797-4158-83aa-0134c8b1a933)
- Upon investigation, the GCP logs revealed the following error: [Link to GCP logs](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/assets/30013840/fc7f8431-4797-4158-83aa-0134c8b1a933)
- The connection is established using GORM with `cloud-sql-go-connector`.
- The application itself is a Go language service built on the Echo framework.

**Error Description:**
The error seems to stem from the database connection being rejected after updating the Go dependency. Despite the connection being established through `cloud-sql-go-connector`, the logs indicate an issue with maintaining connectivity.

**Steps to Reproduce:**
1. Update the Go dependency.
2. Attempt to establish a connection to the database using `cloud-sql-go-connector`.
3. Observe the rejection of the database connection.

**Expected Behavior:**
The application should be able to establish and maintain a connection to the database using `cloud-sql-go-connector` after updating the Go dependency.

**Environment:**
- Application: Go language service
- Framework: Echo
- Database Connector: `cloud-sql-go-connector`

**Additional Information:**
- Any relevant information or context that could assist in troubleshooting or understanding the issue.

**Screenshots/Logs:**
![image](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/assets/30013840/0cfcfcc5-9519-4afc-8a58-e6fda32f8a94)
- Screenshot of the error message or relevant logs: [Link to screenshot/log](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/assets/30013840/0cfcfcc5-9519-4afc-8a58-e6fda32f8a94)
- Any other pertinent visual or textual information related to the issue.

**Priority:**
This issue is affecting the functionality of the application and requires urgent attention.

**Assignees:**
Assign appropriate individuals or teams responsible for addressing this issue.

**Labels:**
Apply relevant labels to categorize and prioritize this issue.

---

Feel free to customize this draft further based on your specific needs or additional information you may have.

### Example code (or command)

```Go
func New(pg_url, gcpCredentialDbbase64 string) *gorm.DB {

	gcpCredentialDb, err := base64.StdEncoding.DecodeString(gcpCredentialDbbase64)
	if err != nil {
		log.Fatal(err)
	}

	cleanup, err := pgxv4.RegisterDriver("cloudsql-postgres", cloudsqlconn.WithCredentialsJSON(gcpCredentialDb))
	if err != nil {
		log.Fatal(err)
	}
	defer cleanup()

	db, err := gorm.Open(postgres.New(postgres.Config{
		DriverName: "cloudsql-postgres",
		DSN:        pg_url,
	}))

	if err != nil {
		log.Println(err)
		os.Exit(1)
	}
	return db
}
```

### Steps to reproduce?

1. ?
2. ?
3. ?
...

### Environment

1. OS type and version:
2. Go version:
3. Cloud SQL Go Connector version:

**Comment by jackwotherspoon** [comment]
Hi @jSierraB3991 Thanks for raising an issue on the Cloud SQL Go Connector 😄 

What is actually happening here is that you have a small bug in your code that was previously going unnoticed in older versions and not being properly raised. We just fixed it as of v1.9.0 via https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/pull/766

Let me explain what is going on here and the source of the error: 

This error resoles around the `cleanup`. Let's take a look at the  `RegisterDriver` code to understand what is happening: 
https://github.com/GoogleCloudPlatform/cloud-sql-go-connector/blob/c4fb69f9b7ec32be3bd07fbf399b5dc1ca0da88c/postgres/pgxv4/postgres.go#L31-L46

The key piece here is how `RegisterDriver` returns `d.Close()` which in your code sample is your `cleanup` variable.

When you use `defer cleanup()` you are essentially saying once your `New()` func  returns the db object, immediately call cleanup which is `d.Close()` to close the `Dialer`. You need the dialer to be open to establish future connections and refresh any previous connections. 

So in your code sample you should `return db, cleanup` and move your `defer cleanup()` to whereever you are calling your `New()` func from.

This is why we added the nice error to point this out. Previously as you mention it wouldn't error immediately but would error once the single connection expired (not making it obvious the true cause of the error) which is why you may not have run into it. 

We try to make this clear in our [README sample](https://github.com/GoogleCloudPlatform/cloud-sql-go-connector?tab=readme-ov-file#using-the-dialer-with-databasesql) but we can probably update it to be even more clear:
```go
// call cleanup when you're done with the database connection
defer cleanup()
```

This was originally caught in the AlloyDB mirror of this library: https://github.com/GoogleCloudPlatform/alloydb-go-connector/issues/522#issuecomment-2018490297

Hope this clears up the confusion. If you have any follow-up questions or need any more clarity please ask them here 😄

**Comment by jSierraB3991** [comment]
@jackwotherspoon Thanks
